File: 단계별로 풀어보기/백트래킹/15649/N과 M (1).cpp
#include <iostream>
#define MAX 8
using namespace std;

int N, M;
int arr[MAX] = {0,};
bool visited[MAX] = {0,};

int dfs(int cnt){
    if(cnt == M){
        for(int i = 0; i < M; i++){
            cout << arr[i]+1 << ' ';
        }
        
        cout << '\n';
    }
    
    for(int i = 0; i < N; i++){
        if(!visited[i]){
            visited[i] = true;
            arr[cnt]=i;
            dfs(cnt+1);
            visited[i] = false;
        }
    }
    
    return 0;
}



int main(){
    cin >> N >> M;
    dfs(0);
    return 0;
}

File: 단계별로 풀어보기/백트래킹/15649/N과 M (1).py
import sys

def dfs(cnt):
    
    if cnt == M:
        for i in range(M):
            sys.stdout.write('%d ' %arr[i])
        sys.stdout.write('\n')
        
    for i in range(N):
        if not visited[i]:
            visited[i] = True
            arr[cnt] = i+1
            dfs(cnt+1)
            visited[i] = False

N, M = map(int, sys.stdin.readline().split())
arr = [0 for i in range(8)]
visited = [False for i in range(8)]
dfs(0)
